lib/widgets/sing_in_buttons.dart

```diff
@@ -0,0 +1,180 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io' show Platform;
+import 'dart:math';
+import 'package:crypto/crypto.dart';
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:google_sign_in/google_sign_in.dart';
+import 'package:sign_in_with_apple/sign_in_with_apple.dart' as apple;
+
+class SignInButtons extends StatefulWidget {
+  const SignInButtons({
+    this.onSignOut,
+    this.onSingIn,
+    Key? key,
+  }) : super(key: key);
+
+  final FutureOr<void> Function()? onSingIn;
+  final FutureOr<void> Function()? onSignOut;
+
+  @override
+  _SignInButtonsState createState() => _SignInButtonsState();
+}
+
+class _SignInButtonsState extends State<SignInButtons> {
+  var loading = false;
+
+  Future<void> signIn(
+    Future<AuthCredential?> Function() getCredentials,
+  ) async {
+    await signOut();
+    setState(() {
+      loading = true;
+    });
+    final credentials = await getCredentials();
+    if (credentials != null) {
+      await FirebaseAuth.instance.signInWithCredential(credentials);
+      await widget.onSingIn?.call();
+    }
+    setState(() {
+      loading = false;
+    });
+  }
+
+  Future<void> signOut() async {
+    setState(() {
+      loading = true;
+    });
+    final provider =
+        FirebaseAuth.instance.currentUser?.providerData.first.providerId;
+    if (provider == null) return;
+
+    await FirebaseAuth.instance.signOut();
+    switch (provider) {
+      case 'google.com':
+        await GoogleSignIn().signOut();
+        break;
+      case 'apple.com':
+        // TODO await appleSignOut;
```

Comment:
Is `FirebaseAuth.instance.signOut()` enough for full correct sign out, or is there additional Apple-specific step like that with Google?

Reply:
I couldn't find an additional one, but It logs me out yes, so I suppose it's enough